server/src/main/java/org/apache/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -219,82 +219,111 @@ public void stop()
-                  payload -> {
-                    try {
-                      return jsonMapper.readValue(payload, DataSegment.class);
-                    }
-                    catch (IOException e) {
-                      throw new RuntimeException(e);
-                    }
-                  }
-              )
-
-          )
-      );
+      return enableSegments(dataSource, FOREVER);
+    }
+    catch (Exception e) {
+      log.error(e, "Exception enabling datasource %s", dataSource);
+      return false;
+    }
+  }
 
-      final List<DataSegment> segments = new ArrayList<>();
-      List<TimelineObjectHolder<String, DataSegment>> timelineObjectHolders = segmentTimeline.lookup(
-          Intervals.of("0000-01-01/3000-01-01")
-      );
-      for (TimelineObjectHolder<String, DataSegment> objectHolder : timelineObjectHolders) {
-        for (PartitionChunk<DataSegment> partitionChunk : objectHolder.getObject()) {
-          segments.add(partitionChunk.getObject());
-        }
-      }
+  @Override
+  public boolean enableSegments(final String dataSource, final Interval interval)
+  {
+    VersionedIntervalTimeline<String, DataSegment> versionedIntervalTimeline = getVersionedIntervalTimeline(dataSource);
 
-      if (segments.isEmpty()) {
-        log.warn("No segments found in the database!");
-        return false;
-      }
+    return enableSegments(
+        segmentIdsForInterval(versionedIntervalTimeline, interval).collect(Collectors.toSet()),
+        versionedIntervalTimeline
+    );
+  }
 
-      dbi.withHandle(
-          new HandleCallback<Void>()
-          {
-            @Override
-            public Void withHandle(Handle handle)
-            {
-              Batch batch = handle.createBatch();
+  @Override
+  public boolean enableSegments(final String dataSource, final Collection<String> segmentIds)
+  {
+    VersionedIntervalTimeline<String, DataSegment> versionedIntervalTimeline = getVersionedIntervalTimeline(dataSource);
 
-              for (DataSegment segment : segments) {
-                batch.add(
-                    StringUtils.format(
-                        "UPDATE %s SET used=true WHERE id = '%s'",
-                        getSegmentsTable(),
-                        segment.getId()
-                    )
-                );
-              }
-              batch.execute();
+    return enableSegments(
+        segmentIdsForInterval(versionedIntervalTimeline, FOREVER)
+            .filter(segmentId -> segmentIds.contains(segmentId.toString()))
+            .collect(Collectors.toSet()),
+        versionedIntervalTimeline
+    );
+  }
 
-              return null;
-            }
-          }
-      );
-    }
-    catch (Exception e) {
-      log.error(e, "Exception enabling datasource %s", dataSource);
+  private boolean enableSegments(
+      final Collection<SegmentId> segmentIds,
+      final VersionedIntervalTimeline<String, DataSegment> versionedIntervalTimeline
+  )
+  {
+    if (segmentIds.isEmpty()) {
```

Comment:
I think it's okay to keep the warning log if no segments found for the interval in the db